src/py/mat3ra/made/tools/build/defect/builders.py

```diff
@@ -173,23 +188,124 @@ def create_adatom(
     def get_equidistant_position(
         self, material: Material, position_on_surface: List[float], distance_z: float = 2.0
     ) -> List[float]:
-        new_basis = material.basis
-        adatom_position = self._calculate_position_from_2d(material, position_on_surface, distance_z)
-        neighboring_atoms_ids = get_nearest_neighbors_atom_indices(material, adatom_position)
+        new_basis = material.basis.copy()
+        adatom_coordinate = self._calculate_coordinate_from_position_and_distance(
+            material, position_on_surface, distance_z
+        )
+        neighboring_atoms_ids = get_nearest_neighbors_atom_indices(material, adatom_coordinate)
         # We need to check if neighboring atoms number is the same in pbc
         supercell_material = create_supercell(material, [[3, 0, 0], [0, 3, 0], [0, 0, 1]])
         # Move the coordinate to the central unit cell of the supercell (crystal coordinates)
-        supercell_adatom_position = [1 / 3 + adatom_position[0] / 3, 1 / 3 + adatom_position[1] / 3, adatom_position[2]]
+        supercell_adatom_coordinate = [
+            1 / 3 + adatom_coordinate[0] / 3,
+            1 / 3 + adatom_coordinate[1] / 3,
+            adatom_coordinate[2],
+        ]
         supercell_neighboring_atoms_ids = get_nearest_neighbors_atom_indices(
-            supercell_material, supercell_adatom_position
+            supercell_material, supercell_adatom_coordinate
         )
         if neighboring_atoms_ids is None or supercell_neighboring_atoms_ids is None:
             raise ValueError("No neighboring atoms found. Try reducing the distance_z.")
         if len(supercell_neighboring_atoms_ids) != len(neighboring_atoms_ids):
             raise ValueError("Number of neighboring atoms is not the same in PBC. Try increasing the supercell size.")
-        neighboring_atoms_coordinates = [new_basis.coordinates.values[atom_id] for atom_id in neighboring_atoms_ids]
+        new_basis.coordinates.filter_by_ids(neighboring_atoms_ids)
```

Comment:
Let's identify the nearest neighbors in a supercell and fold the resulting coordinate back